CI: Kani verifier v0.49.0 uses Rust `nightly-2024-03-29`, which cannot compile `proc-macro2@1.0.80`

As of April 16, 2024, the latest version v0.49.0 of Kani verifier uses Rust `nightly-2024-03-29`, which cannot compile `proc-macro2@1.0.80`, causing a CI job to tail.

https://github.com/moka-rs/moka/actions/runs/8691269712/job/23852829412

```console
...
Installed Kani  successfully
[0/5] Running Kani first-time setup...
[1/5] Ensuring the existence of: /home/runner/.kani/kani-0.49.0
[2/5] Downloading Kani release bundle: kani-0.49.0-x86_64-unknown-linux-gnu.tar.gz
[3/5] Installing rust toolchain version: nightly-2024-03-29-x86_64-unknown-linux-gnu
info: syncing channel updates for 'nightly-2024-03-29-x86_64-unknown-linux-gnu'
info: latest update on 2024-03-29, rust version 1.79.0-nightly (ba527200c 2024-03-28)
...
  nightly-2024-03-29-x86_64-unknown-linux-gnu installed - rustc 1.79.0-nightly (ba527200c 2024-03-28)

...
cargo-kani --features 'sync, future'

rror[E0658]: use of unstable library feature 'proc_macro_byte_character'
   --> /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/proc-macro2-1.0.80/src/wrapper.rs:871:21
    |
871 |                     proc_macro::Literal::byte_character(byte)
    |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: see issue #115268 <https://github.com/rust-lang/rust/issues/115268> for more information
    = help: add `#![feature(proc_macro_byte_character)]` to the crate attributes to enable
    = note: this compiler was built on 2024-03-28; consider upgrading it if it is out of date

error[E0658]: use of unstable library feature 'proc_macro_c_str_literals'
   --> /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/proc-macro2-1.0.80/src/wrapper.rs:898:21
    |
898 |                     proc_macro::Literal::c_string(string)
    |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: see issue #119750 <https://github.com/rust-lang/rust/issues/119750> for more information
    = help: add `#![feature(proc_macro_c_str_literals)]` to the crate attributes to enable
    = note: this compiler was built on 2024-03-28; consider upgrading it if it is out of date

error: aborting due to 2 previous errors
```

--- a/.ci_extras/pin-crate-vers-kani.sh
+++ b/.ci_extras/pin-crate-vers-kani.sh
@@ -1,0 +1,8 @@
+#!/bin/sh
+
+set -eux
+
+# Pin some dependencies to specific versions for the nightly toolchain
+# used by Kani verifier.
+cargo update -p syn@2.0 --precise 2.0.58
+cargo update -p proc-macro2 --precise 1.0.79

--- a/.github/workflows/Kani.yml
+++ b/.github/workflows/Kani.yml
@@ -43,6 +43,12 @@ jobs:
           lscpu
           free -m
 
+      - name: Pin some dependencies to specific versions
+        run: ./.ci_extras/pin-crate-vers-kani.sh
+
+      - name: Show cargo tree
+        run: cargo tree
+
       - name: Run Kani
         uses: model-checking/kani-github-action@v1.0
         with:

